File: scripts/e2e/qc-import.py
#!/usr/bin/env python3
"""Fallback QC Excel import when npm `xlsx` is unavailable. Writes qc/catalog.json."""
from __future__ import annotations

import argparse
import json
import os
import re
import sys
from pathlib import Path

try:
    import openpyxl
except ImportError:
    import subprocess

    subprocess.check_call([sys.executable, "-m", "pip", "install", "openpyxl", "-q"])
    import openpyxl

KIT_ROOT = Path(__file__).resolve().parents[2]
SKIP = {
    "Cover",
    "Guideline",
    "Revision History",
    "Summary",
    "Dashboard",
    "Report Test",
    "Bug Data",
    "RTM",
}


def refuse_base() -> None:
    if KIT_ROOT.name == "project-quality-kit" and os.environ.get("QUALITY_ALLOW_BASE_INIT") != "1":
        print(
            f"[qc:import.py] Refusing to mutate Base template '{KIT_ROOT}'. "
            "Clone to <project>-quality first. Maintainer override: QUALITY_ALLOW_BASE_INIT=1.",
            file=sys.stderr,
        )
        sys.exit(1)


def load_cfg():
    p = KIT_ROOT / "_meta" / "project.json"
    if p.exists():
        return json.loads(p.read_text(encoding="utf-8"))
    yml = KIT_ROOT / "_meta" / "project.yml"
    if yml.exists():
        try:
            import yaml  # type: ignore

            return yaml.safe_load(yml.read_text(encoding="utf-8")) or {}
        except Exception:
            return {}
    return {}


def norm_header(cell) -> str:
    """Normalize Excel header: newlines / pipes / trailing (hint) → stable key."""
    s = str(cell or "")
    s = s.replace("\n", " ").replace("|", " ")
    s = re.sub(r"\s+", " ", s).strip().lower()
    s = re.sub(r"\s*\([^)]*\)\s*$", "", s).strip()
    return s


def col(fmap, row, *names):
    for n in names:
        i = fmap.get(n.lower())
        if i is not None and i < len(row) and row[i] is not None:
            return str(row[i]).strip()
    return ""


def main():
    refuse_base()
    ap = argparse.ArgumentParser()
    ap.add_argument("--file", default="")
    ap.add_argument("--priority", default="")
    ap.add_argument("--group", default="")
    ap.add_argument("--sheet", default="")
    args = ap.parse_args()

    cfg = load_cfg()
    file = args.file
    if not file:
        input_dir = KIT_ROOT / "qc" / "input"
        files = sorted(input_dir.glob("*.xlsx")) if input_dir.exists() else []
        files = [f for f in files if not f.name.startswith("~$")]
        if not files:
            print("[qc:import.py] Provide --file or place xlsx in qc/input/", file=sys.stderr)
            sys.exit(1)
        file = str(files[0])
    path = Path(file).resolve()
    if not path.exists():
        print("[qc:import.py] File not found:", path, file=sys.stderr)
        sys.exit(1)

    wb = openpyxl.load_workbook(path, read_only=True, data_only=True)
    cases = []
    for sheet_name in wb.sheetnames:
        if sheet_name in SKIP:
            continue
        if args.sheet and sheet_name != args.sheet:
            continue
        rows = list(wb[sheet_name].iter_rows(values_only=True))
        header_idx = None
        fmap = {}
        for i, row in enumerate(rows[:20]):
            cells = [str(c) if c is not None else "" for c in (row or ())]
            if any(re.search(r"testcase\s*id", c, re.I) for c in cells):
                header_idx = i
                for idx, c in enumerate(row or ()):
                    key = norm_header(c)
                    if key:
                        fmap[key] = idx
                break
        if header_idx is None:
            continue
        for row in rows[header_idx + 1 :]:
            row = row or ()
            id_ = col(fmap, row, "testcase id")
            if not id_ or not id_.startswith("TC_"):
                continue
            priority = col(fmap, row, "priority")
            group = col(fmap, row, "group")
            if args.priority and args.priority.lower() not in priority.lower():
                continue
            if args.group and group.lower() != args.group.lower():
                continue
            cases.append(
                {
                    "id": id_,
                    "reqId": col(fmap, row, "req id"),
                    "docSource": col(fmap, row, "doc source"),
                    "group": group,
                    "priority": priority,
                    "title": col(fmap, row, "test title"),
                    "precondition": col(fmap, row, "pre-condition"),
                    "steps": col(fmap, row, "test steps"),
                    "expected": col(fmap, row, "expected result"),
                    "origin": col(fmap, row, "origin"),
                    "sheet": sheet_name,
                    "automated": col(fmap, row, "automated"),
                    "script": col(fmap, row, "script"),
                }
            )
    wb.close()

    out_rel = (cfg.get("qc") or {}).get("catalog_out") or "qc/catalog.json"
    out = KIT_ROOT / out_rel
    out.parent.mkdir(parents=True, exist_ok=True)
    from datetime import datetime, timezone

    payload = {
        "source": str(path),
        "importedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "count": len(cases),
        "cases": cases,
    }
    out.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"[qc:import.py] {len(cases)} case(s) → {out}")
    from collections import Counter

    print("[qc:import.py] by priority:", dict(Counter(c["priority"] for c in cases)))
    print("[qc:import.py] by group:", dict(Counter(c["group"] for c in cases)))


if __name__ == "__main__":
    main()

File: scripts/e2e/qc-import.mjs
#!/usr/bin/env node
/**
 * Import ISC QC Excel → qc/catalog.json
 * VI: Đọc sheet chức năng (bỏ Cover/Guideline/Summary/…), lấy cột Testcase ID + metadata.
 * EN: Reads feature sheets, extracts Testcase ID + metadata into a catalog.
 *
 * Usage:
 *   npm run qc:import -- --file ../path/to/ISC_*_TestCase.xlsx
 *   npm run qc:import -- --file qc/input/cases.xlsx --priority High
 */
import fs from "node:fs";
import path from "node:path";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import { assertNotBaseTemplate } from "../lib/refuse-base.mjs";
import { loadProjectCfg } from "../lib/load-project-cfg.mjs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const KIT_ROOT = path.resolve(__dirname, "../..");
assertNotBaseTemplate(KIT_ROOT, "qc:import");

const require = createRequire(import.meta.url);
const XLSX = require("xlsx");

const SKIP_SHEETS = new Set([
  "Cover",
  "Guideline",
  "Revision History",
  "Summary",
  "Dashboard",
  "Report Test",
  "Bug Data",
  "RTM",
]);

function parseArgs(argv) {
  const out = { file: "", priority: "", group: "", sheet: "" };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--file") out.file = argv[++i];
    else if (a === "--priority") out.priority = argv[++i];
    else if (a === "--group") out.group = argv[++i];
    else if (a === "--sheet") out.sheet = argv[++i];
  }
  return out;
}

function loadCfg() {
  return loadProjectCfg(KIT_ROOT);
}

function normHeader(cell) {
  let s = String(cell ?? "")
    .replace(/\n/g, " ")
    .replace(/\|/g, " ")
    .replace(/\s+/g, " ")
    .trim()
    .toLowerCase();
  s = s.replace(/\s*\([^)]*\)\s*$/, "").trim();
  return s;
}

function headerIndex(row) {
  const map = {};
  row.forEach((cell, idx) => {
    if (cell == null) return;
    const key = normHeader(cell);
    if (key) map[key] = idx;
  });
  return map;
}

function col(map, row, ...names) {
  for (const n of names) {
    const idx = map[n.toLowerCase()];
    if (idx != null && row[idx] != null) return String(row[idx]).trim();
  }
  return "";
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const cfg = loadCfg();
  let file = args.file;
  if (!file) {
    const inputDir = path.join(KIT_ROOT, "qc/input");
    const files = fs.existsSync(inputDir)
      ? fs.readdirSync(inputDir).filter((f) => f.endsWith(".xlsx") && !f.startsWith("~"))
      : [];
    if (!files.length) {
      console.error("[qc:import] Provide --file <xlsx> or place a file in qc/input/");
      process.exit(1);
    }
    file = path.join(inputDir, files[0]);
  }
  file = path.resolve(file);
  if (!fs.existsSync(file)) {
    console.error("[qc:import] File not found:", file);
    process.exit(1);
  }

  const wb = XLSX.readFile(file, { cellDates: true });
  const cases = [];

  for (const sheetName of wb.SheetNames) {
    if (SKIP_SHEETS.has(sheetName)) continue;
    if (args.sheet && sheetName !== args.sheet) continue;
    const ws = wb.Sheets[sheetName];
    const rows = XLSX.utils.sheet_to_json(ws, { header: 1, defval: null, raw: false });
    // Find header row containing "Testcase ID"
    let headerRowIdx = -1;
    let map = {};
    for (let i = 0; i < Math.min(rows.length, 20); i++) {
      const joined = (rows[i] || []).map((c) => String(c ?? "")).join(" | ");
      if (/testcase\s*id/i.test(joined)) {
        headerRowIdx = i;
        map = headerIndex(rows[i]);
        break;
      }
    }
    if (headerRowIdx < 0) continue;

    for (let r = headerRowIdx + 1; r < rows.length; r++) {
      const row = rows[r] || [];
      const id = col(map, row, "testcase id");
      if (!id || !/^TC_/i.test(id)) continue;
      const priority = col(map, row, "priority");
      const group = col(map, row, "group");
      if (args.priority && !priority.toLowerCase().includes(args.priority.toLowerCase())) continue;
      if (args.group && group.toLowerCase() !== args.group.toLowerCase()) continue;

      cases.push({
        id,
        reqId: col(map, row, "req id"),
        docSource: col(map, row, "doc source"),
        group,
        priority,
        title: col(map, row, "test title"),
        precondition: col(map, row, "pre-condition"),
        steps: col(map, row, "test steps"),
        expected: col(map, row, "expected result"),
        origin: col(map, row, "origin"),
        sheet: sheetName,
        automated: col(map, row, "automated"),
        script: col(map, row, "script"),
      });
    }
  }

  const outRel = cfg.qc?.catalog_out || "qc/catalog.json";
  const out = path.resolve(KIT_ROOT, outRel);
  fs.mkdirSync(path.dirname(out), { recursive: true });
  const payload = {
    source: file,
    importedAt: new Date().toISOString(),
    count: cases.length,
    cases,
  };
  fs.writeFileSync(out, JSON.stringify(payload, null, 2));
  console.log(`[qc:import] ${cases.length} case(s) → ${out}`);
  const byPri = cases.reduce((acc, c) => {
    const k = c.priority || "?";
    acc[k] = (acc[k] || 0) + 1;
    return acc;
  }, {});
  const byGroup = cases.reduce((acc, c) => {
    const k = c.group || "?";
    acc[k] = (acc[k] || 0) + 1;
    return acc;
  }, {});
  console.log("[qc:import] by priority:", byPri);
  console.log("[qc:import] by group:", byGroup);
}

main();
